client/tso_request_dispatcher.go

```diff
@@ -0,0 +1,751 @@
+		dcLocation := dcLocationKey.(string)
+		// Skip the Global TSO Allocator
+		if dcLocation == globalDCLocation {
+			return true
+		}
+		if dispatcher, exist := c.bc.GetTSOAllocators().Load(dcLocation); !exist {
+			log.Info("[pd] delete unused tso dispatcher", zap.String("dc-location", dcLocation))
+			dispatcher.(*tsoDispatcher).dispatcherCancel()
+			c.tsoDispatcher.Delete(dcLocation)
+		}
+		return true
+	})
+}
+
+type deadline struct {
+	timer  <-chan time.Time
+	done   chan struct{}
+	cancel context.CancelFunc
+}
+
+func (c *client) tsCancelLoop() {
+	defer c.wg.Done()
+
+	tsCancelLoopCtx, tsCancelLoopCancel := context.WithCancel(c.ctx)
+	defer tsCancelLoopCancel()
+
+	ticker := time.NewTicker(tsLoopDCCheckInterval)
+	defer ticker.Stop()
+	for {
+		// Watch every dc-location's tsDeadlineCh
+		c.bc.GetTSOAllocators().Range(func(dcLocation, _ interface{}) bool {
+			c.watchTSDeadline(tsCancelLoopCtx, dcLocation.(string))
+			return true
+		})
+		select {
+		case <-c.checkTSDeadlineCh:
+			continue
+		case <-ticker.C:
+			continue
+		case <-tsCancelLoopCtx.Done():
+			return
+		}
+	}
+}
+
+func (c *client) watchTSDeadline(ctx context.Context, dcLocation string) {
+	if _, exist := c.tsDeadline.Load(dcLocation); !exist {
+		tsDeadlineCh := make(chan deadline, 1)
+		c.tsDeadline.Store(dcLocation, tsDeadlineCh)
+		go func(dc string, tsDeadlineCh <-chan deadline) {
+			for {
+				select {
+				case d := <-tsDeadlineCh:
+					select {
+					case <-d.timer:
+						log.Error("[pd] tso request is canceled due to timeout", zap.String("dc-location", dc), errs.ZapError(errs.ErrClientGetTSOTimeout))
+						d.cancel()
+					case <-d.done:
+						continue
+					case <-ctx.Done():
+						return
+					}
+				case <-ctx.Done():
+					return
+				}
+			}
+		}(dcLocation, tsDeadlineCh)
+	}
+}
+
+func (c *client) scheduleCheckTSDeadline() {
+	select {
+	case c.checkTSDeadlineCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *client) tsLoop() {
+	defer c.wg.Done()
+
+	loopCtx, loopCancel := context.WithCancel(c.ctx)
+	defer loopCancel()
+
+	ticker := time.NewTicker(tsLoopDCCheckInterval)
+	defer ticker.Stop()
+	for {
+		c.updateTSODispatcher()
+		select {
+		case <-ticker.C:
+		case <-c.checkTSODispatcherCh:
+		case <-loopCtx.Done():
+			return
+		}
+	}
+}
+
+func (c *client) checkAllocator(
+	dispatcherCtx context.Context,
+	forwardCancel context.CancelFunc,
+	dc, forwardedHostTrim, addrTrim, url string,
+	updateAndClear func(newAddr string, connectionCtx *tsoConnectionContext)) {
+	defer func() {
+		// cancel the forward stream
+		forwardCancel()
+		requestForwarded.WithLabelValues(forwardedHostTrim, addrTrim).Set(0)
+	}()
+	cc, u := c.bc.GetTSOAllocatorClientConnByDCLocation(dc)
+	healthCli := healthpb.NewHealthClient(cc)
+	for {
+		// the pd/allocator leader change, we need to re-establish the stream
+		if u != url {
+			log.Info("[pd] the leader of the allocator leader is changed", zap.String("dc", dc), zap.String("origin", url), zap.String("new", u))
```

Comment:
also [pd]?

Reply:
These functions are used by both tso APIs in pd client and tso APIs in the independent tso client. I changed name to [pd/tso]

Reply:
Also changed every [pd] to [pd/tso] in this file.